gems/aws-sdk-cognitoidentity/lib/aws-sdk-cognitoidentity/customizations/cognito_identity_credentials.rb

```diff
@@ -0,0 +1,132 @@
+    # the Amazon Cognito Identity service.
+    #
+    # This provider gets credentials using the
+    # {Client.get_credentials_for_identity} service operation, which
+    # requires either an `identity_id` or an `identity_pool_id` (Amazon Cognito
+    # Identity Pool ID), which is used to call {Client.get_id} to
+    # obtain an `identity_id` automatically.
+    #
+    # In addition, if this credential provider is used to provide authenticated
+    # login, the `logins` map may be set to the tokens provided by the
+    # respective identity providers. See {constructor} for an example on
+    # creating a credentials object with proper property values.
+    #
+    # ## Refreshing Credentials from Identity Service
+    #
+    # the CognitoIdentityCredentials will auto-refresh the AWS credentials from
+    # Cognito.  In addition to AWS credentials expiring after a given amount of
+    # time, the login token from the identity provider will also expire.
+    # Once this token expires, it will not be usable to refresh AWS credentials,
+    # and another token will be needed. The SDK does not manage refreshing of
+    # the token value, but this can be done through a "refresh token"
+    #  supported by most identity providers. Consult the documentation for
+    # the identity provider for refreshing tokens. Once the refreshed token is
+    # acquired, you should make sure to update this new token in the
+    # CognitoIdentityCredentials object's {logins} property. The following
+    # code will update the WebIdentityToken, assuming you have retrieved
+    # an updated token from the identity provider:
+    #
+    #     AWS.config.credentials.logins['graph.facebook.com'] = updatedToken;
+    #     AWS.config.credentials.refresh!  # required only if authentication state has changed
+    #
+    # The CognitoIdentityCredentials also provides a `before_refresh` callback
+    # that can be used to help manage refreshing identity provider tokens.
+    # `before_refresh` is called when AWS credentials are required and need
+    # to be refreshed and it has access to the CognitoIdentityCredentials object.
+    class CognitoIdentityCredentials
+
+      include CredentialProvider
+      include RefreshingCredentials
+
+      # @param [Hash] options
+      # @option options [String] :identity_id the Cognito identity_id.  Required
+      #   unless identity_pool_id is given. A unique
+      #   identifier in the format REGION:GUID
+      #
+      # @option options [String] :identity_pool_id Required unless identity_id
+      #   is provided. A Amazon Cognito
+      #   Identity Pool ID)in the format REGION:GUID.
+      #
+      # @option options [Hash<String,String>] :logins A set of optional
+      #   name-value pairs that map provider names to provider tokens.
+      #   The name-value pair will follow the syntax
+      #   "provider_name": "provider_user_identifier".
+      #
+      # @option options [String] :custom_role_arn (optional) The Amazon Resource
```

Comment:
in Yard, `(optional)` will actually lay out a default of "optional". If you want to say something is required, I believe it's `[required, String]` Otherwise, if not required, it's assumed optional.